### PR TITLE
[WIP][DTensor] Add a util to compute DTensor padding

### PR DESCRIPTION
**Pull request by wz337**
Currently, we are doing dynamic padding on DTensor, but we want to explore static sharding for performance purposes, which means we want to do padding before any sharding happens. 

This PR to add a a util to compute the padding needed on each tensor dim on a global level. 

For example, let's say we have:
```
world_size = 6
mesh_shape = (3, 2)
placements = (Shard(0), Shard(0))
tensor = ([0, 1, 2, 3, 4, 5, 6])
```

With the current dynamic padding, we have the following sharding result:
```
tensor: [0, 1, 2, 3, 4, 5, 6]

rank0: [0, 1], rank1: [2]
rank2: [3, 4], rank3: [5]
rank4: [6], rank5: []
```

With static padding, we will decide the padding size needed for each tensor dimension instead of mesh dimension before sharding happens. 
```
# This PR calculates the padding required on the tensor dimension based on mesh_shape and placements.
# Padding is [5], which means we need to pad 5 elements on the 0th dimension of the given tensor.
tensor: [0, 1, 2, 3, 4, 5, 6, #, #, #, #, #]

rank0: [0, 1], rank1: [2, 3]
rank2: [4, 5], rank3: [6, #]
rank4: [#, #], rank5: [#, #]
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_tensor/_utils.py [hunk 1]
```diff
@@ -224,3 +224,28 @@ def compute_local_stride(
     return tuple(
         global_stride[i] // stride_divisors[i] for i in range(len(global_stride))
     )
+
+
+def compute_global_padding(tensor: "dtensor.DTensor") -> List[int]:
```
Comment (Skylion007) [inline torch/distributed/_tensor/_utils.py:229]: just import properly behind an if `TYPE_CHECKING` flag

File: torch/distributed/_tensor/_utils.py [hunk 2]
```diff
@@ -224,3 +224,30 @@ def compute_local_stride(
...
+                num_shard_by_dim[tensor_shard_dim] = mesh_dim_size
+            else:
+                num_shard_by_dim[tensor_shard_dim] *= mesh_dim_size
+
+    pad_sizes = [0 for _ in range(len(global_shape))]
+    for tensor_dim, num_shard in enumerate(num_shard_by_dim):
+        tensor_dim_size = global_shape[tensor_dim]
+        padded_tensor_dim_size = (tensor_dim_size + num_shard) - (
+            tensor_dim_size % num_shard
+        )
```
Comment (XilunWu) [inline torch/distributed/_tensor/_utils.py:250]: what is the `padded_tensor_dim_size` if `num_shard == ` e.g. the placement is replicate().

File: test/distributed/_tensor/test_utils.py [hunk 3]
```diff
@@ -122,6 +123,17 @@ def test_compute_local_shape_and_global_offset_2D(self):
...
+    @with_comms
+    def test_compute_global_padding(self):
+        local_tensor = torch.randn(8, requires_grad=True)
+        mesh = init_device_mesh(self.device_type, (3, 2))
+        dist_tensor = distribute_tensor(local_tensor, mesh, [Shard(0), Shard(0)])
+        padding = compute_global_padding(
+            dist_tensor.shape, dist_tensor.device_mesh, dist_tensor.placements
+        )
+        self.assertEqual(len(padding), dist_tensor.ndim)
+        self.assertEqual(padding, [4])
```
Comment (XilunWu) [inline test/distributed/_tensor/test_utils.py:135]: n00b question. Why should it be 4? My understanding of this test is:
shard a tensor of size [8] into 6 ranks (device mesh: `[[0, 1], [2, 3], [4, 5]]`)by `chunk()`, the first 4 shards are of size [2] and the other 2 shards are empty. Shouldn't the padding be [0] on rank 0, 1, 2, 3; and [2] on rank 4 and 5?

BTW, what is the expected result of sharding size [9] tensor into this 3x2 mesh? The padded size should still be 2 on all ranks. Which ranks will have padding? My question is mainly about the order of putting first unpadded tensors when sharding one tensor dim over multiple mesh dims.

**Review by XilunWu — COMMENTED** [review]
have some questions.

File: torch/distributed/_tensor/_utils.py (hunk 2 above)
Comment (wz337) [reply torch/distributed/_tensor/_utils.py]: Ah. Good catch. Fixed by skipping num_shard_by_dim == -1

File: test/distributed/_tensor/test_utils.py (hunk 3 above)
Comment (wz337) [reply test/distributed/_tensor/test_utils.py]: Good question. I added some summary in the PR to show what we are trying to do. Basically, we want to explore doing padding before sharding. See if the summary gives a better idea and we can discuss this offline as well.

**Comment by wanchaol** [comment]
The summary should be "dynamic padding" instead of "dynamic sharding"

File: torch/distributed/_tensor/_utils.py [hunk 4]
```diff
@@ -224,3 +224,31 @@ def compute_local_stride(
...
     )
+
+
+def compute_global_padding(
+    global_shape: ShapeType, mesh: DeviceMesh, placements: Sequence[Placement]
+) -> List[int]:
+    """
+    Compute the list of padding needed to make the tensor evenly shardable.
+    """
+    num_shard_by_dim = [-1 for _ in range(len(global_shape))]
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:235]: the default should be "1" instead of "-1"? "1" means replicated bascially

File: torch/distributed/_tensor/_utils.py [hunk 5]
```diff
@@ -224,3 +224,31 @@ def compute_local_stride(
...
+    for mesh_idx, placement in enumerate(placements):
+        if placement.is_shard():
+            tensor_shard_dim = placement.dim  # type: ignore[attr-defined]
+            mesh_dim_size = mesh.size(mesh_idx)
+            if num_shard_by_dim[tensor_shard_dim] == -1:
+                num_shard_by_dim[tensor_shard_dim] = mesh_dim_size
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:241]: making default value be "1" would unify this if-else

File: test/distributed/_tensor/test_utils.py [hunk 6]
```diff
@@ -122,6 +123,17 @@ def test_compute_local_shape_and_global_offset_2D(self):
                     global_tensor[dim0_start:dim0_end, dim1_start:dim1_end],
                 )
 
+    @with_comms
+    def test_compute_global_padding(self):
+        local_tensor = torch.randn(7, 8, requires_grad=True)
+        mesh = init_device_mesh(self.device_type, (3, 2))
+        dist_tensor = distribute_tensor(local_tensor, mesh, [Shard(0), Shard(0)])
```
Comment (wanchaol) [inline test/distributed/_tensor/test_utils.py:130]: in this test case, can you test various sharding combinations instead of only (Shard(0), Shard(0))? including the case where the num_shards exceed the tensor dim size

File: torch/distributed/_tensor/_utils.py [hunk 7]
```diff
@@ -224,3 +224,31 @@ def compute_local_stride(
     return tuple(
         global_stride[i] // stride_divisors[i] for i in range(len(global_stride))
     )
+
+
+def compute_global_padding(
+    global_shape: ShapeType, mesh: DeviceMesh, placements: Sequence[Placement]
+) -> List[int]:
+    """
+    Compute the list of padding needed to make the tensor evenly shardable.
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:233]: please give more precise comments: what's the meaning of the return List[int] value.

File: torch/distributed/_tensor/_utils.py [hunk 8]
```diff
@@ -224,3 +224,31 @@ def compute_local_stride(
...
+            if num_shard_by_dim[tensor_shard_dim] == -1:
+                num_shard_by_dim[tensor_shard_dim] = mesh_dim_size
+            else:
+                num_shard_by_dim[tensor_shard_dim] *= mesh_dim_size
+
+    pad_sizes = [0 for _ in range(len(global_shape))]
+    for tensor_dim, num_shard in enumerate(num_shard_by_dim):
+        if num_shard != -1:
+            tensor_dim_size = global_shape[tensor_dim]
+            padded_tensor_dim_size = (tensor_dim_size + num_shard) - (
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:249]: this calculation this kinda hard to understand.. can you use the existing formula in the current dynamic padding case?  https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/placement_types.py#L95

**Review by wanchaol — COMMENTED** [review]
Please see inline comments
